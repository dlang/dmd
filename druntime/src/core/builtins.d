/**********************************************
To provide access to features that would be otherwise counterproductive or
difficult to implement, compilers provide an interface consisting of a set
of builtins (also called intrinsics) which can be called like normal functions.

This module exposes builtins both common to all D compilers
(those provided by the frontend) and specific to the host compiler i.e. those
specific to either LLVM or GCC (`ldc.intrinsics` and `gcc.builtins` are publicly imported, respectively).
Host-specific intrinsics cannot be reliably listed here, however listings can be found
at the documentation for the relevant backends, i.e.
$(LINK2 https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html, GCC) and
$(LINK2 https://llvm.org/docs/LangRef.html, LLVM). It should be noted that not all
builtins listed are necessarily supported by the host compiler, please file a bug
if this is the case for your workload.

Use of this module reduces the amount of conditional compilation needed
to use a given builtin. For example, to write a target independent function
that uses prefetching we can write the following:
---
float usePrefetch(float[] x)
{
    // There is only one import statement required rather than two (versioned) imports
    import core.builtins;
    version (GNU)
        __builtin_prefetch(x.ptr);
    version (LDC)
        /+
            For the curious: 0, 3, 1 mean `x` will only be read-from (0), it will be used
            very often (3), and it should be fetched to the data-cache (1).
        +/
        llvm_prefetch(x.ptr, 0, 3, 1);
    const doMath = blahBlahBlah;
    return doMath;
}
---


Copyright: Copyright Â© 2021, The D Language Foundation
License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
Authors:   Walter Bright
Source:    $(DRUNTIMESRC core/builtins.d)
*/

module core.builtins;

version (GNU)
    public import gcc.builtins;

version (LDC)
    public import ldc.intrinsics;

/// Writes `s` to `stderr` during CTFE (does nothing at runtime).
void __ctfeWrite(scope const(char)[] s) @nogc @safe pure nothrow {}

version (GNU)
{
    /// https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fexpect
    alias expect = __builtin_expect;
    /// https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005ftrap
    alias trap = __builtin_trap;
}
else version (LDC)
{
    /// https://llvm.org/docs/LangRef.html#llvm-expect-intrinsic
    alias expect = llvm_expect;
    debug
        /// https://llvm.org/docs/LangRef.html#llvm-debugtrap-intrinsic
        alias trap = llvm_debugtrap;
    else
        /// https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic
        alias trap = llvm_trap;
}
else version (DigitalMars)
{
    pragma(inline, true)
    T expect(T)(T val, T expected) if (__traits(isIntegral, T))
    {
        return val;
    }

    /// Execute target dependent trap instruction, if supported.
    /// Otherwise, abort execution.
    pragma(inline, true)
    void trap()
    {
        debug
        {
            version(D_InlineAsm_X86)
                asm nothrow @nogc pure @trusted { int 3; }
        }
        assert(0);
    }
}

/// Provide static branch and value hints for the LDC/GDC compilers.
/// DMD ignores these hints.
pragma(inline, true) bool likely()(bool b)   { return !!expect(b, true);  }
/// ditto
pragma(inline, true) bool unlikely()(bool b) { return !!expect(b, false); }

///
@nogc nothrow pure @safe unittest
{
    int x = 12;

    expect(x, 12);

    if (likely(x > 0))
    {
        // ...
    }
    else if (unlikely(x == int.min))
    {
        // ...
    }
}
