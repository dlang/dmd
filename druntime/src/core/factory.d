/* Create classes from their modules and names.
 *
 * Copyright: Copyright (C) D Language Foundation 2023
 * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
 * Authors:   Walter Bright, Steven Schveighoffer
 * Source:    $(DRUNTIMESRC core/_factory.d)
 */

module core.factory;

/**
 * Create instance of class specified by the module symbol and a string
 * representing the name of the class.
 * The class must either have no constructors or have
 * a default constructor.
 * Params:
 *   mod = symbol representing the module that the class is in
 *   classname = string representing the name of the class
 * Returns:
 *   null if failed
 * Example:
 * ---
 * module foo.bar;
 *
 * class C
 * {
 *     this() { x = 10; }
 *     int x;
 * }
 *
 * void main()
 * {
 *     auto c = cast(C)factory!(foo.bar)("C");
 *     assert(c !is null && c.x == 10);
 * }
 * ---
 */
Object factory(alias mod)(string classname)
{
    foreach(cl; _getModuleClasses!mod)
    {
        if (cl.stringof == classname)
            return cl.classinfo.create();
    }
    return null;
}

@system unittest
{
    Object valid_obj = factory!object("Object");
    Object invalid_obj = factory!object("__this_class_doesnt_exist__");

    assert(valid_obj !is null);
    assert(invalid_obj is null);
}

/**************************************
 * Retrieve as a tuple all the types of the top level classes in the module mod.
 */
private template _getModuleClasses(alias mod) {
   alias result = _AliasSeq!();
   static foreach(m; __traits(allMembers, mod))
      static if(is(__traits(getMember, mod, m) == class))
         result = _AliasSeq!(result, __traits(getMember, mod, m));
   alias _getModuleClasses = result;
}

private template _AliasSeq(TList...) { alias _AliasSeq = TList; }
