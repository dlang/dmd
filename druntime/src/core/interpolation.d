/++
    This module provides definitions to support D's
    interpolated expression sequence literal, sometimes
    called string interpolation.


    ---
    string str;
    int num;
    // the compiler uses this module to implement the
    // i"..." literal used here.
    auto a = i"$​(str) has $​(num) items.";
    ---

    The variable `a` is a sequence of expressions:

    ---
    a[0] == InterpolationHeader()
    a[$-1] == InterpolationFooter()
    ---

    First and last, you see the header and footer, to
    clearly indicate where interpolation begins and ends.
    Note that there may be nested interpolated sequences too,
    each with their own header and footer. Think of them
    as a set of balanced parenthesis around the contents.

    Inside, you will find three general categories of
    content: `InterpolatedLiteral!"string"` for string
    expressions, `InterpolatedExpression!"code"` for code
    expressions, and then the values themselves as their
    own type.

    In the example:
    ---
    auto a = i"$​(str) has $​(num) items.";
    ---

    We will find:
    ---
    a[0] == InterpolationHeader()
    a[1] == InterpolatedExpression!"str"
    a[2] == str
    a[3] == InterpolatedLiteral!" has ";
    a[4] == InterpolatedExpression!"num";
    a[5] == num
    a[6] == InterpolatedLiteral!" items.";
    a[7] == InterpolationFooter()
    a.length == 8;
    ---

    You can see the correspondence with the original
    input: when you write `$​(expression)`, the string of the
    expression is passed as `InterpolatedExpression!ThatString`,
    (excluding any parenthesis around the expression),
    and everything else is passed as `InterpolatedLiteral!str`,
    in the same sequence as they appeared in the source.

    After an `InterpolatedExpression!...`, you will find the
    actual value(s) in the tuple. (If the expression expanded
    to multiple values - for example, if it was itself a tuple,
    there will be multiple values for a single expression.)

    Library functions should NOT attempt to mixin the code
    from an `InterpolatedExpression` themselves. Doing so
    will fail, since it is coming from a different scope anyway.
    The string is provided to you only for informational purposes
    and as a sentinel to separate things the user wrote.

    Your code should be able to handle an empty code string
    in `InterpolatedExpression` or even an entirely missing
    `InterpolatedExpression`, in case an implementation decides to
    not emit these.

    The `toString` members on these return `null`, except for
    the `InterpolatedLiteral`, which returns the literal string.
    This is to ease processing by generic functions like
    `std.stdio.write` or `std.conv.text`, making them effectively
    transparently skipped.

    To extract the string from an `InterpolatedLiteral`, you can
    use an `is` expression or the `.toString` method.

    To extract the string from a `InterpolatedExpression`, you can
    use an `is` expression or the `.expression` member.

    None of these structures have runtime state.

    History:
        Added in dmd 2.10x frontend, released in late 2023.
+/
module core.interpolation;

/++
    Common implementation for returning an empty string, to avoid storing
    multiple versions of the same function based on templated types below.
+/
public string __getEmptyString() @nogc pure nothrow @safe {
    return "";
}

/++
    Sentinel values to indicate the beginning and end of an
    interpolated expression sequence.

    Note that these can nest, so while processing a sequence,
    it may be helpful to keep a nesting count if that knowledge
    is important to your application.
+/
struct InterpolationHeader {
    /++
        Returns `null` for easy compatibility with existing functions
        like `std.stdio.writeln` and `std.conv.text`.
    +/
    alias toString = __getEmptyString;
}

/// ditto
struct InterpolationFooter {
    /++
        Returns `null` for easy compatibility with existing functions
        like `std.stdio.writeln` and `std.conv.text`.
    +/
    alias toString = __getEmptyString;
}

/++
    Represents a fragment of a string literal in between expressions
    passed as part of an interpolated expression sequence.
+/
struct InterpolatedLiteral(string text) {
    /++
        Returns the text of the interpolated string literal for this
        segment of the tuple, for easy access and compatibility with
        existing functions like `std.stdio.writeln` and `std.conv.text`.
    +/
    static string toString() @nogc pure nothrow @safe {
        return text;
    }
}

/++
    Represents the source code of an expression passed as part of an
    interpolated expression sequence.
+/
struct InterpolatedExpression(string text) {
    /++
        Returns the text of an interpolated expression used in the
        original literal, if provided by the implementation.
    +/
    enum expression = text;

    /++
        Returns `null` for easy compatibility with existing functions
        like `std.stdio.writeln` and `std.conv.text`.
    +/
    alias toString = __getEmptyString;
}
