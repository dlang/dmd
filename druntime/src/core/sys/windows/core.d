/**
 * Helper module for the Windows API
 *
 * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
 * Source: $(DRUNTIMESRC core/sys/windows/_core.d)
 */
module core.sys.windows.core;
version (Windows):

/**
 The core Windows API functions.

 Importing this file is equivalent to the C code:
 ---
 #define WIN32_LEAN_AND_MEAN
 #include "windows.h"
 ---

*/

public import core.sys.windows.windef;
public import core.sys.windows.winnt;
public import core.sys.windows.wincon;
public import core.sys.windows.winbase;
public import core.sys.windows.wingdi;
public import core.sys.windows.winuser;
public import core.sys.windows.winnls;
public import core.sys.windows.winver;
public import core.sys.windows.winnetwk;
public import core.sys.windows.winsvc;
