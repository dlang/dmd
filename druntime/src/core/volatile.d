/**
 * This module declares intrinsics for volatile operations.
 *
 * Copyright: Copyright Â© 2019, The D Language Foundation
 * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
 * Authors:   Walter Bright, Ernesto Castellotti
 * Source:    $(DRUNTIMESRC core/volatile.d)
 */

module core.volatile;

nothrow:
@safe:
@nogc:

/*************************************
 * Read/write value from/to the memory location indicated by ptr.
 *
 * These functions are recognized by the compiler, and calls to them are guaranteed
 * to not be removed (as dead assignment elimination or presumed to have no effect)
 * or reordered in the same thread.
 *
 * These reordering guarantees are only made with regards to other
 * operations done through these functions; the compiler is free to reorder regular
 * loads/stores with regards to loads/stores done through these functions.
 *
 * This is useful when dealing with memory-mapped I/O (MMIO) where a store can
 * have an effect other than just writing a value, or where sequential loads
 * with no intervening stores can retrieve
 * different values from the same location due to external stores to the location.
 *
 * These functions will, when possible, do the load/store as a single operation. In
 * general, this is possible when the size of the operation is less than or equal to
 * $(D (void*).sizeof), although some targets may support larger operations. If the
 * load/store cannot be done as a single operation, multiple smaller operations will be used.
 *
 * These are not to be conflated with atomic operations. They do not guarantee any
 * atomicity. This may be provided by coincidence as a result of the instructions
 * used on the target, but this should not be relied on for portable programs.
 * Further, no memory fences are implied by these functions.
 * They should not be used for communication between threads.
 * They may be used to guarantee a write or read cycle occurs at a specified address.
 */

ubyte  volatileLoad(ubyte * ptr);
ushort volatileLoad(ushort* ptr);  /// ditto
uint   volatileLoad(uint  * ptr);  /// ditto
ulong  volatileLoad(ulong * ptr);  /// ditto

void volatileStore(ubyte * ptr, ubyte  value);   /// ditto
void volatileStore(ushort* ptr, ushort value);   /// ditto
void volatileStore(uint  * ptr, uint   value);   /// ditto
void volatileStore(ulong * ptr, ulong  value);   /// ditto

@system unittest
{
    alias TT(T...) = T;

    foreach (T; TT!(ubyte, ushort, uint, ulong))
    {
        T u;
        T* p = &u;
        volatileStore(p, 1);
        T r = volatileLoad(p);
        assert(r == u);
    }
}
