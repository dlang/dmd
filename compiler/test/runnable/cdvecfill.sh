#!/usr/bin/env bash

export LC_ALL=C
src_file=${EXTRA_FILES}/${TEST_NAME}.d
expect_file=${EXTRA_FILES}/${TEST_NAME}.out
expect_file_avx=${EXTRA_FILES}/${TEST_NAME}avx.out
expect_file_avx2=${EXTRA_FILES}/${TEST_NAME}avx2.out
tmp_file=${RESULTS_TEST_DIR}/${TEST_NAME}.out
obj_file=${RESULTS_TEST_DIR}/${TEST_NAME}.o

if [ $OS == "linux" ] && [ $MODEL == "64" ]; then
  $DMD -betterC -c -O -m64 ${src_file} -of${obj_file}
  objdump --disassemble --disassembler-options=intel ${obj_file} | tail -n+3 | sed 's/[ \t]\s*$//' > ${tmp_file}
  diff ${expect_file} ${tmp_file}
  rm_retry ${obj_file} ${tmp_file}

  $DMD -betterC -c -O -m64 -mcpu=avx ${src_file} -of${obj_file}
  objdump --disassemble --disassembler-options=intel ${obj_file} | tail -n+3 | sed 's/[ \t]\s*$//' > ${tmp_file}
  diff ${expect_file_avx} ${tmp_file}
  rm_retry ${obj_file} ${tmp_file}

  $DMD -betterC -c -O -m64 -mcpu=avx2 ${src_file} -of${obj_file}
  objdump --disassemble --disassembler-options=intel ${obj_file} | tail -n+3 | sed 's/[ \t]\s*$//' > ${tmp_file}
  diff ${expect_file_avx2} ${tmp_file}
  rm_retry ${obj_file} ${tmp_file}
fi
