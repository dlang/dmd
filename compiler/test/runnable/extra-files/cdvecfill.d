import core.simd;

version (D_AVX)
{
    __vector(ubyte[16]) test(ubyte val) { return val; }
    __vector(ubyte[16]) test(ubyte* pval) { return *pval; }
    __vector(byte[16]) test(byte val) { return val; }
    __vector(byte[16]) test(byte* pval) { return *pval; }

    __vector(ushort[8]) test(ushort val) { return val; }
    __vector(ushort[8]) test(ushort* pval) { return *pval; }
    __vector(short[8]) test(short val) { return val; }
    __vector(short[8]) test(short* pval) { return *pval; }

    __vector(uint[4]) test(uint val) { return val; }
    __vector(uint[4]) test(uint* pval) { return *pval; }
    __vector(int[4]) test(int val) { return val; }
    __vector(int[4]) test(int* pval) { return *pval; }

    __vector(ulong[2]) test(ulong val) { return val; }
    __vector(ulong[2]) test(ulong* pval) { return *pval; }
    __vector(long[2]) test(long val) { return val; }
    __vector(long[2]) test(long* pval) { return *pval; }

    __vector(float[4]) test(float val) { return val; }
    __vector(float[4]) test(float* pval) { return *pval; }
    __vector(double[2]) test(double val) { return val; }
    __vector(double[2]) test(double* pval) { return *pval; }

    __vector(ubyte[32]) test2(ubyte val) { return val; }
    __vector(ubyte[32]) test2(ubyte* pval) { return *pval; }
    __vector(byte[32]) test2(byte val) { return val; }
    __vector(byte[32]) test2(byte* pval) { return *pval; }

    __vector(ushort[16]) test2(ushort val) { return val; }
    __vector(ushort[16]) test2(ushort* pval) { return *pval; }
    __vector(short[16]) test2(short val) { return val; }
    __vector(short[16]) test2(short* pval) { return *pval; }

    __vector(uint[8]) test2(uint val) { return val; }
    __vector(uint[8]) test2(uint* pval) { return *pval; }
    __vector(int[8]) test2(int val) { return val; }
    __vector(int[8]) test2(int* pval) { return *pval; }

    __vector(ulong[4]) test2(ulong val) { return val; }
    __vector(ulong[4]) test2(ulong* pval) { return *pval; }
    __vector(long[4]) test2(long val) { return val; }
    __vector(long[4]) test2(long* pval) { return *pval; }

    __vector(float[8]) test2(float val) { return val; }
    __vector(float[8]) test2(float* pval) { return *pval; }
    __vector(double[4]) test2(double val) { return val; }
    __vector(double[4]) test2(double* pval) { return *pval; }
}
else version (D_AVX)
{
    __vector(ubyte[16]) test(ubyte val) { return val; }
    __vector(ubyte[16]) test(ubyte* pval) { return *pval; }
    __vector(byte[16]) test(byte val) { return val; }
    __vector(byte[16]) test(byte* pval) { return *pval; }

    __vector(ushort[8]) test(ushort val) { return val; }
    __vector(ushort[8]) test(ushort* pval) { return *pval; }
    __vector(short[8]) test(short val) { return val; }
    __vector(short[8]) test(short* pval) { return *pval; }

    __vector(uint[4]) test(uint val) { return val; }
    __vector(uint[4]) test(uint* pval) { return *pval; }
    __vector(int[4]) test(int val) { return val; }
    __vector(int[4]) test(int* pval) { return *pval; }

    __vector(ulong[2]) test(ulong val) { return val; }
    __vector(ulong[2]) test(ulong* pval) { return *pval; }
    __vector(long[2]) test(long val) { return val; }
    __vector(long[2]) test(long* pval) { return *pval; }

    __vector(float[4]) test(float val) { return val; }
    __vector(float[4]) test(float* pval) { return *pval; }
    __vector(double[2]) test(double val) { return val; }
    __vector(double[2]) test(double* pval) { return *pval; }

    __vector(ubyte[32]) test2(ubyte val) { return val; }
    __vector(ubyte[32]) test2(ubyte* pval) { return *pval; }
    __vector(byte[32]) test2(byte val) { return val; }
    __vector(byte[32]) test2(byte* pval) { return *pval; }

    __vector(ushort[16]) test2(ushort val) { return val; }
    __vector(ushort[16]) test2(ushort* pval) { return *pval; }
    __vector(short[16]) test2(short val) { return val; }
    __vector(short[16]) test2(short* pval) { return *pval; }

    __vector(uint[8]) test2(uint val) { return val; }
    __vector(uint[8]) test2(uint* pval) { return *pval; }
    __vector(int[8]) test2(int val) { return val; }
    __vector(int[8]) test2(int* pval) { return *pval; }

    __vector(ulong[4]) test2(ulong val) { return val; }
    __vector(ulong[4]) test2(ulong* pval) { return *pval; }
    __vector(long[4]) test2(long val) { return val; }
    __vector(long[4]) test2(long* pval) { return *pval; }

    __vector(float[8]) test2(float val) { return val; }
    __vector(float[8]) test2(float* pval) { return *pval; }
    __vector(double[4]) test2(double val) { return val; }
    __vector(double[4]) test2(double* pval) { return *pval; }
}
else
{
    __vector(ubyte[16]) test(ubyte val) { return val; }
    __vector(ubyte[16]) test(ubyte* pval) { return *pval; }
    __vector(byte[16]) test(byte val) { return val; }
    __vector(byte[16]) test(byte* pval) { return *pval; }

    __vector(ushort[8]) test(ushort val) { return val; }
    __vector(ushort[8]) test(ushort* pval) { return *pval; }
    __vector(short[8]) test(short val) { return val; }
    __vector(short[8]) test(short* pval) { return *pval; }

    __vector(uint[4]) test(uint val) { return val; }
    __vector(uint[4]) test(uint* pval) { return *pval; }
    __vector(int[4]) test(int val) { return val; }
    __vector(int[4]) test(int* pval) { return *pval; }

    __vector(ulong[2]) test(ulong val) { return val; }
    __vector(ulong[2]) test(ulong* pval) { return *pval; }
    __vector(long[2]) test(long val) { return val; }
    __vector(long[2]) test(long* pval) { return *pval; }

    __vector(float[4]) test(float val) { return val; }
    __vector(float[4]) test(float* pval) { return *pval; }
    __vector(double[2]) test(double val) { return val; }
    __vector(double[2]) test(double* pval) { return *pval; }
}
