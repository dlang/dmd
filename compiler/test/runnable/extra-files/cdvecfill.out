

Disassembly of section .text._D9cdvecfill4testFhZNhG16h:

0000000000000000 <_D9cdvecfill4testFhZNhG16h>:
   0:	66 0f 6e c7          	movd   xmm0,edi
   4:	66 0f 60 c0          	punpcklbw xmm0,xmm0
   8:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   c:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
  11:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPhZNhG16h:

0000000000000000 <_D9cdvecfill4testFPhZNhG16h>:
   0:	0f b6 07             	movzx  eax,BYTE PTR [rdi]
   3:	66 0f 6e c0          	movd   xmm0,eax
   7:	66 0f 60 c0          	punpcklbw xmm0,xmm0
   b:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   f:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
  14:	c3                   	ret
  15:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFgZNhG16g:

0000000000000000 <_D9cdvecfill4testFgZNhG16g>:
   0:	66 0f 6e c7          	movd   xmm0,edi
   4:	66 0f 60 c0          	punpcklbw xmm0,xmm0
   8:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   c:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
  11:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPgZNhG16g:

0000000000000000 <_D9cdvecfill4testFPgZNhG16g>:
   0:	0f be 07             	movsx  eax,BYTE PTR [rdi]
   3:	66 0f 6e c0          	movd   xmm0,eax
   7:	66 0f 60 c0          	punpcklbw xmm0,xmm0
   b:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   f:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
  14:	c3                   	ret
  15:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFtZNhG8t:

0000000000000000 <_D9cdvecfill4testFtZNhG8t>:
   0:	66 0f 6e c7          	movd   xmm0,edi
   4:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   8:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
   d:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPtZNhG8t:

0000000000000000 <_D9cdvecfill4testFPtZNhG8t>:
   0:	0f b7 07             	movzx  eax,WORD PTR [rdi]
   3:	66 0f 6e c0          	movd   xmm0,eax
   7:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   b:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
  10:	c3                   	ret
  11:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFsZNhG8s:

0000000000000000 <_D9cdvecfill4testFsZNhG8s>:
   0:	66 0f 6e c7          	movd   xmm0,edi
   4:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   8:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
   d:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPsZNhG8s:

0000000000000000 <_D9cdvecfill4testFPsZNhG8s>:
   0:	0f bf 07             	movsx  eax,WORD PTR [rdi]
   3:	66 0f 6e c0          	movd   xmm0,eax
   7:	66 0f 61 c0          	punpcklwd xmm0,xmm0
   b:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
  10:	c3                   	ret
  11:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFkZNhG4k:

0000000000000000 <_D9cdvecfill4testFkZNhG4k>:
   0:	66 0f 6e c7          	movd   xmm0,edi
   4:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
   9:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPkZNhG4k:

0000000000000000 <_D9cdvecfill4testFPkZNhG4k>:
   0:	66 0f 6e 07          	movd   xmm0,DWORD PTR [rdi]
   4:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
   9:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFiZNhG4i:

0000000000000000 <_D9cdvecfill4testFiZNhG4i>:
   0:	66 0f 6e c7          	movd   xmm0,edi
   4:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
   9:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPiZNhG4i:

0000000000000000 <_D9cdvecfill4testFPiZNhG4i>:
   0:	66 0f 6e 07          	movd   xmm0,DWORD PTR [rdi]
   4:	66 0f 70 c0 00       	pshufd xmm0,xmm0,0x0
   9:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFmZNhG2m:

0000000000000000 <_D9cdvecfill4testFmZNhG2m>:
   0:	66 48 0f 6e c7       	movq   xmm0,rdi
   5:	66 0f 6c c0          	punpcklqdq xmm0,xmm0
   9:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPmZNhG2m:

0000000000000000 <_D9cdvecfill4testFPmZNhG2m>:
   0:	66 0f 6c 07          	punpcklqdq xmm0,XMMWORD PTR [rdi]
   4:	c3                   	ret
   5:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFlZNhG2l:

0000000000000000 <_D9cdvecfill4testFlZNhG2l>:
   0:	66 48 0f 6e c7       	movq   xmm0,rdi
   5:	66 0f 6c c0          	punpcklqdq xmm0,xmm0
   9:	c3                   	ret
	...

Disassembly of section .text._D9cdvecfill4testFPlZNhG2l:

0000000000000000 <_D9cdvecfill4testFPlZNhG2l>:
   0:	66 0f 6c 07          	punpcklqdq xmm0,XMMWORD PTR [rdi]
   4:	c3                   	ret
   5:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFfZNhG4f:

0000000000000000 <_D9cdvecfill4testFfZNhG4f>:
   0:	50                   	push   rax
   1:	f3 0f 11 04 24       	movss  DWORD PTR [rsp],xmm0
   6:	f3 0f 10 04 24       	movss  xmm0,DWORD PTR [rsp]
   b:	0f c6 c0 00          	shufps xmm0,xmm0,0x0
   f:	59                   	pop    rcx
  10:	c3                   	ret
  11:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFPfZNhG4f:

0000000000000000 <_D9cdvecfill4testFPfZNhG4f>:
   0:	f3 0f 10 07          	movss  xmm0,DWORD PTR [rdi]
   4:	0f c6 c0 00          	shufps xmm0,xmm0,0x0
   8:	c3                   	ret
   9:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFdZNhG2d:

0000000000000000 <_D9cdvecfill4testFdZNhG2d>:
   0:	50                   	push   rax
   1:	f2 0f 11 04 24       	movsd  QWORD PTR [rsp],xmm0
   6:	f2 0f 10 04 24       	movsd  xmm0,QWORD PTR [rsp]
   b:	66 0f 14 c0          	unpcklpd xmm0,xmm0
   f:	59                   	pop    rcx
  10:	c3                   	ret
  11:	00 00                	add    BYTE PTR [rax],al
	...

Disassembly of section .text._D9cdvecfill4testFPdZNhG2d:

0000000000000000 <_D9cdvecfill4testFPdZNhG2d>:
   0:	f2 0f 10 07          	movsd  xmm0,QWORD PTR [rdi]
   4:	66 0f 14 c0          	unpcklpd xmm0,xmm0
   8:	c3                   	ret
   9:	00 00                	add    BYTE PTR [rax],al
	...
